decisions/0003-cf-api-controller-repo.md

```diff
@@ -0,0 +1,89 @@
+# 3. CF API Controller Source Code Organization
+
+Date: 2020-06-17
+
+## Status
+
+In Review
+
+## Context
+
+Currently we store `capi-kpack-watcher` directly in `capi-k8s-release`. This has
+caused some complexity when bumping image versions in our concourse pipeline.
+We're also planning on reworking the watcher to use [kubebuilder](https://github.com/kubernetes-sigs/kubebuilder), and so now is a
+good time to change any source code organization decisions.
+
+Proposed locations for `capi-kpack-watcher` source code:
+
+1. `cf-api-kpack-watcher` a repo just for this particular component
+
+   1. Pros:
+
+      1. Very clear what goes where
+
+      1. It'd be a standard `kubebuilder` go repo
+
+      1. Fits our current pattern of go service repositories (1 repo for each service)
+
+   1. Cons:
+
+      1. Many more repos to manage to keep this pattern consistent in the
+       near-future as we will also create controllers for routes, droplets, etc
+
+1. `cf-api-controllers`/`cf-controllers` which would contain other future controllers we create for CF API resources
+
+   1. Pros:
+
+      1. It'd be a standard `kubebuilder` go repo with multiple controllers
+
+      1. Only one more repo to manage to keep this pattern consistent
+
+   1. Cons:
+
+      1. Possibly unclear in the future whether controllers without CF API dependency should go here. (This is specifically
+       if we use a more generic name like `cf-controllers` which doesn't make clear this repository is only managed by our team)
+
+      1. Marrying the versions of all these controllers together makes it
+         difficult if you ever to want use different effective versions of
+         source for the controllers
+          - e.g. if we introduce a bug in `controller-a` source, but want to
+            keep the same version of `controller-b` source
+
+1. `cloud_controller_ng` mono-repo
+
+   1. Pros:
+
+      1. 0 more repos to manage to keep this pattern consistent
+
+      1. No way for API to get out-of-sync with controllers - they all come from
+       the same source code.
+          - Versioning, for instance, becomes simplified a bit with this
+
+   1. Cons:
+
+      1. 1 repo with Ruby and Go code inside can be _very_ annoying to grep.
```

Comment:
I'm not sure this is really that bad. `ag` for example has flags for searching only in specific file types